src/Orleans/GrainDirectory/IGrainDirectory.cs

```diff
@@ -0,0 +1,67 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans.GrainDirectory
+{
+
+    /// <summary>
+    /// Recursive distributed operations on grain directories.
+    /// Each operation may forward the request to a remote owner, increasing the hopcount.
+    /// </summary>
+    interface IGrainDirectory
+    {
+        /// <summary>
+        /// Record a new grain activation by adding it to the directory.
+        /// <para>This method must be called from a scheduler thread.</para>
+        /// </summary>
+        /// <param name="address">The address of the new activation.</param>
+        /// <param name="singleact">If true, use single-activation registration</param>
+        /// <param name="withRetry">Indicates whether or not to retry the operation.</param>
+        /// <param name="hopcount">Counts recursion depth across silos</param>
+        /// <returns>The registered address and the version associated with this directory mapping.</returns>
+        Task<Tuple<ActivationAddress, int>> RegisterAsync(ActivationAddress address, bool singleact, int hopcount = 0);
```

Comment:
Struct would be fine with me. I think it also has better performance than Tuple (which is an object).